src/components/shared/MarkerTooltipContents.js

```diff
@@ -37,19 +39,80 @@ function _markerDetail<T: NotVoidOrNull>(
   fn: T => string = String
 ): React.Node {
   return [
-    <div className="tooltipLabel" key="{key}">
+    <div className="tooltipLabel" key={key}>
       {label}:
     </div>,
     fn(value),
   ];
 }
 
+type PhaseTimeTuple = { name: string, time: Microseconds };
+
+function _markerDetailPhase(p: PhaseTimeTuple): React.Node {
+  return _markerDetail(
+    'gcphase' + p.name,
+    'Phase ' + p.name,
+    p.time,
+    t => formatNumber(t / 1000) + 'ms'
+  );
+}
+
+function _makePhaseTimesArray(
+  phases: PhaseTimes<Microseconds>
+): Array<PhaseTimeTuple> {
+  const array = [];
+  for (const phase in phases) {
+    if (phase !== 'Total') {
+      array.push({ name: phase, time: phases[phase] });
+    }
+  }
+  return array;
+}
+
+function _leafPhase(phases: Array<PhaseTimeTuple>, name: string): boolean {
+  for (let i = 0; i < phases.length; i++) {
+    const pt = phases[i];
+    if (pt.name === name) {
+      // Don't consider itself.
+      continue;
+    }
+    if (pt.name.startsWith(name)) {
+      // Name is not a leaf
+      return false;
+    }
+  }
+  // name is a leaf.
+  return true;
+}
+
+function _filterInterestingPhaseTimes(
+  raw_phases: PhaseTimes<Microseconds>,
+  num_select: number
+): Array<PhaseTimeTuple> {
+  let phase_times = _makePhaseTimesArray(raw_phases);
+
+  // Select only the leaf phases.
+  phase_times = phase_times.filter(pt => _leafPhase(phase_times, pt.name));
+
+  // Find the top N phases.
+  const threshold = Math.min(
+    ...phase_times
+      .map(pt => pt.time)
+      .sort((a, b) => Number(b) - Number(a))
+      .slice(0, num_select + 1)
```

Comment:
why `+1` ? If we `slice` between `0` and `num_select` we get `num_select` elements, isn't it what we want?

Reply:
ah, I think you use "+1" because otherwise you didn't get enough elements... because you used `>` in the filtering predicate, while you wanted `>=` :)
(but also read the comment below)